nx/lib/nx/serving.ex

```diff
@@ -176,17 +252,329 @@ defmodule Nx.Serving do
+      |> Enum.zip_with(&Nx.concatenate(&1, axis: @axis))
+
+    {output, []} =
+      Nx.Defn.Composite.traverse(template, tensors, fn _template, [tensor | tensors] ->
+        {tensor, tensors}
+      end)
+
+    {output, metadata}
+  end
+
+  defp receive_batched(total_size, ref, acc, _template_metadata, name, input) do
+    receive do
+      {^ref, {start, size, output, metadata}} ->
+        # If we have a single response, slice and return immediately.
+        # Otherwise we collect their contents and build the concatenated result later.
+        if acc == [] and size == total_size do
+          Process.demonitor(ref, [:flush])
+
+          output =
+            Nx.Defn.Composite.traverse(output, &Nx.slice_along_axis(&1, start, size, axis: @axis))
+
+          {output, metadata}
+        else
+          funs =
+            output
+            |> Nx.Defn.Composite.reduce(
+              [],
+              &[Nx.slice_along_axis(&1, start, size, axis: @axis) | &2]
+            )
+            |> Enum.reverse()
+
+          receive_batched(total_size - size, ref, [funs | acc], {output, metadata}, name, input)
+        end
+
+      {:DOWN, ^ref, _, _, reason} ->
+        # We fake monitor messages, so still demonitor and flush.
+        Process.demonitor(ref, [:flush])
+        exit({reason, {__MODULE__, :batched_run, [name, input]}})
+    end
+  end
+
+  ## Process callbacks
+
+  require Logger
+  @behaviour GenServer
+
+  @empty_stack {[], 0}
+  @empty_queue :queue.new()
+  @timeout_message {__MODULE__, :timeout}
+
+  @doc false
+  def start_child(opts) do
+    {serving, opts} = Keyword.pop!(opts, :serving)
+    {batch_size, opts} = Keyword.pop(opts, :batch_size, 1)
+    {batch_timeout, opts} = Keyword.pop(opts, :batch_timeout, 100)
+    name = Keyword.fetch!(opts, :name)
+    GenServer.start_link(__MODULE__, {self(), name, serving, batch_size, batch_timeout}, opts)
+  end
+
+  @impl true
+  def init({parent, name, serving, batch_size, batch_timeout}) do
+    {:ok, module_state} = handle_init(serving.module, :process, serving.arg)
+
+    :persistent_term.put(
+      persistent_key(name),
+      %{
+        limit: batch_size,
+        preprocessing: serving.client_preprocessing,
+        postprocessing: serving.client_postprocessing
+      }
+    )
+
+    # We keep batches in a stack. Once the stack is full
+    # or it times out, we either execute or enqueue it.
+    state = %{
+      name: name,
+      module: serving.module,
+      module_state: module_state,
+      stack: @empty_stack,
+      limit: batch_size,
+      timeout: batch_timeout,
+      timer: :none,
+      queue: @empty_queue,
+      task: :none,
+      task_supervisor: nil
+    }
+
+    {:ok, state, {:continue, {:task_supervisor, parent}}}
+  end
+
+  @impl true
+  def handle_continue({:task_supervisor, parent}, state) do
+    {:ok, task_supervisor} = Supervisor.start_child(parent, Task.Supervisor)
+    {:noreply, %{state | task_supervisor: task_supervisor}}
+  end
```

Comment:
Why don't we list the task supervisor in the supervisor children right away? To avoid naming it?

Reply:
In a way, yes. We would need to generate names for it, which is fine as atoms, but I can see us supporting stuff like `{:via, ..., ...}` in the future, which would break this down.